### PR TITLE
fix: ssh reconnect

### PR DESCRIPTION
**Pull request by rephus**
Closes: #7617

### Description:

TL;DR it seems trying to run a postgres query with an active ssh connection while this is trying to reconnect, it is failing. 

I also upgraded pg libs

Original error: 

Screenshot from 2023-10-25 13-57-18

I fixed this by removing `reconnect` property from ssh 

Now we get the following (not server -killing error) : 

Screenshot from 2023-10-25 15-32-58

I think I also fixed one issue about PG :  capturing `connection terminated unexpectedly` from pg. This seems harder to replicate,so I'm not 100% I fixed this.

Screenshot from 2023-10-25 15-28-17

Related issues: https://github.com/brianc/node-postgres/issues/2283

### How I tested this 

I was killing my ssh periodically 

> while true; do sleep 3; sudo pkill ssh ; sudo service sshd restart ; sudo service ssh restart;  done

I was running queries periodically from terminal

> while true; do sleep 1 ; curl 'http://localhost:3000/api/v1/saved/3ae94507-60df-4585-b8ac-838e31590821/results ... ; done 

[Screencast from 25-10-23 16:27:12.webm](https://github.com/lightdash/lightdash/assets/1983672/e18e8690-14de-4048-87b5-21f1acee8fc8)

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/warehouses/src/ssh/sshTunnel.ts [hunk 1]
```diff
@@ -39,7 +39,9 @@ export class SshTunnel<T extends CreateWarehouseCredentials> {
                             privateKey:
                                 this.originalCredentials.sshTunnelPrivateKey ||
                                 '',
+                            reconnect: false,
```
Comment (rephus) [inline packages/warehouses/src/ssh/sshTunnel.ts:42]: This is the main change

**Comment by owlas** [comment]
@rephus based on you finding that `reconnect` was the problem, I was checking the `ssh2-promises` library and that reconnect feature is one that was added on top of `ssh2`. I checked the code and I think they're missing a catch in this promise chain: https://github.com/sanketbajoria/ssh2-promise/blob/b05272def4b4ac9c4575fec649d0ede078fbd2f3/src/sshConnection.ts#L315-L335

The author made the same error somewhere else too: https://github.com/sanketbajoria/ssh2-promise/issues/73

I'm wondering if we should drop `ssh2-promise` and use the base `ssh2` library and write our own promise wrappers.

File: packages/backend/src/services/ProjectService/ProjectService.ts [hunk 2]
```diff
@@ -1164,11 +1166,9 @@ export class ProjectService {
                             metricQuery: JSON.stringify(metricQuery),
                             type: warehouseClient.credentials.type,
                         },
-                        async () => {
-                            await sshTunnel.testConnection();
```
Comment (owlas) [inline packages/backend/src/services/ProjectService/ProjectService.ts:1168]: what's the reason for removing the ssh test? Is it needed to open the connection to the tunnel ?

File: packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts [hunk 3]
```diff
@@ -170,12 +170,23 @@ export class PostgresClient<
...
+                    done();
+                    return;
+                }
                 if (err) {
                     reject(err);
                     done();
                     return;
                 }
+
+                client.on('error', (e) => {
```
Comment (owlas) [inline packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts:186]: what does this fix?

File: packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts [hunk 4]
```diff
@@ -170,12 +170,23 @@ export class PostgresClient<
             rows: Record<string, any>[];
         }>((resolve, reject) => {
             pool = new pg.Pool(this.config);
+
             pool.connect((err, client, done) => {
+                if (!client) {
```
Comment (owlas) [inline packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts:175]: when is the client undefined?

**Review by owlas — COMMENTED** [review]
some questions

File: packages/backend/src/services/ProjectService/ProjectService.ts (hunk 2 above)
Comment (rephus) [reply packages/backend/src/services/ProjectService/ProjectService.ts:1168]: I added rhis test last week to mitigate the issue . It is not needed anymore.

Afaik, if reconnect is true, connection is open (reconnecting) but tunnel might be closed. If reconmect is false, connection and tunnels will be closed

File: packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts (hunk 3 above)
Comment (rephus) [reply packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts:186]: I removed it, ignore.

File: packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts (hunk 4 above)
Comment (rephus) [reply packages/warehouses/src/warehouseClients/PostgresWarehouseClient.ts:175]: I upgrade the library and its types. It seems connect can return undentified clients, so I'm capturing this as an error

**Comment by rephus** [comment]
While I was testing, I made some extra changes, like the pg lib  upgrade. If you are not sure about thoes changes, I can revert those and only push the `reconnect: false`

**Comment by owlas** [comment]
The pg.lib upgrade is good I want to understand what the new event handlers are for
